integration-libs/epd-visualization/components/visual-viewer/visual-viewer.service.spec.ts

```diff
@@ -0,0 +1,2800 @@
+  let mockElementRef = new MockElementRef();
+  let mockChangeDetectorRef = new MockChangeDetectorRef();
+
+  let visualViewerService: VisualViewerService;
+
+  let viewportAdded$: Observable<void>;
+
+  let originalTimeout: number;
+  let isUi5BootStrapped: () => boolean;
+  let getCore: () => Core;
+  let containerDiv: HTMLDivElement;
+
+  beforeAll(function () {
+    originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;
+    jasmine.DEFAULT_TIMEOUT_INTERVAL = 30000;
+  });
+
+  afterAll(function () {
+    jasmine.DEFAULT_TIMEOUT_INTERVAL = originalTimeout;
+  });
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      providers: [
+        {
+          provide: EpdVisualizationConfig,
+          useValue: getTestConfig(),
+        },
+        {
+          provide: SceneNodeToProductLookupService,
+          useValue: mockSceneNodeToProductLookupService,
+        },
+        {
+          provide: VisualizationLookupService,
+          useValue: mockVisualizationLookupService,
+        },
+        {
+          provide: ElementRef,
+          useValue: mockElementRef,
+        },
+        {
+          provide: ChangeDetectorRef,
+          useValue: mockChangeDetectorRef,
+        },
+      ],
+    });
+
+    visualViewerService = TestBed.inject(VisualViewerService);
+
+    viewportAdded$ = visualViewerService['viewportAdded$'] as Observable<void>;
+    isUi5BootStrapped = visualViewerService['isUi5BootStrapped'];
+    getCore = visualViewerService['getCore'];
+    containerDiv = mockElementRef.nativeElement as HTMLDivElement;
+  });
+
+  describe('constructor', () => {
+    it('should bootstrap UI5, load libraries, add UIArea and add viewport', (done) => {
+      viewportAdded$.subscribe(() => {
+        expect(isUi5BootStrapped()).toEqual(true);
+        expect(getCore()).toBeTruthy();
+
+        expect(getCore().getLoadedLibraries()).toBeTruthy();
+        [
+          'sap.m',
+          'sap.ui.layout',
+          'sap.ui.vk',
+          'sap.ui.richtexteditor',
+        ].forEach((libraryName) =>
+          expect(getCore().getLoadedLibraries()[libraryName]).toBeTruthy()
+        );
+
+        expect(containerDiv.getAttribute('data-sap-ui-area')).toEqual(
+          visualViewerContainerId
+        );
+
+        done();
+      });
+    });
+  });
+
+  describe('backgroundTopColor', () => {
+    it('should do nothing when value has not changed', () => {
+      visualViewerService['_backgroundTopColor'] = 'red';
+      expect(visualViewerService.backgroundTopColor).toEqual('red');
+      const executeWhenSceneLoadedSpy = spyOn<any>(
+        visualViewerService,
+        'executeWhenSceneLoaded'
+      );
+      visualViewerService.backgroundTopColor = 'red';
+      expect(visualViewerService.backgroundTopColor).toEqual('red');
+      expect(executeWhenSceneLoadedSpy).toHaveBeenCalledTimes(0);
+    });
+
+    it('should call setBackgroundColorTop on viewport (after scene has loaded) when value has changed', () => {
```

Comment:
I have problem to match the test description with the test body. In particular the test body is lacking the assertion whether the method `viewport.setBackgroundColorTop()` was called (and ideally: with which params). For example:
```ts
expect(mockViewport.setBackgroundColorTop).toHaveBeenCalledWith('testCssColor')
```

This might require prerequsites:
- `mockViewport` to include the spied method: 
   ```ts
   const mockViewport = { setBackgroundColorTop: jasmine.createSpy('setBackgroundColorTop') };
   ```
- `getCSSColor` method to return mocked value:
   ```ts
   const getCSSColorSpy = spyOn<any>(visualViewerService, 'getCSSColor').and.returnValue('testCssColor');
   ```

If only we want to test more (but I'm not forcing it), we might also want to assert that the `getCSSColor()` method was invoked with proper arguments, for example:
```ts
expect(getCSSColorSpy).toHaveBeenCalledWith('--cx-color-inverse')
```

Reply:
I have added some parameter checking in this test.